### ISSUE TITLE
Slow rate and packet loss in old version of openwrt

### ISSUE THREAD

**Issue by small-5** [ntop/n2n#463]
The latest commit has problems of slow speed and packet loss in the old version of openwrt (kernel version 3.14.79, uclibc), as well as commit 644ea6c82ab3671a6b9aed4494e07b38a823308b. Tried to change the encryption method, but did not solve it. It felt like it was caused by MTU. I didn't know how to find out the cause. There was no problem with the [commit 73fa50a37dc23a91a130335583494c97a2355a95](https://github.com/ntop/n2n/commit/73fa50a37dc23a91a130135583494c97a2355a95) a long time ago, but there were too many changes submitted in the middle. I couldn't find out what the change caused. I hope it can be fixed,Thank you!

**Comment by Logan007** [comment]
If you compiled yourself, would a binary search towards the breaking commit be an option for you? That could really help to look into this.

If you suspect MTU, you might want to play around with edge's `-M` cli parameter.

**Comment by small-5** [comment]
I have checked the MTU of the virtual network card, which is the default 1290. But the symptoms are similar to the loss of large packets, because if I just ping the virtual IP, there will be basically no packet loss, but when there is data transmission, It will start to lose packets, and the speed is significantly slower than the previous version.

**Comment by Logan007** [comment]
Does `sudo ping -f <IP address>` also show packet loss?

And what about different packet sizes `sudo ping -f -s <size> <IP address>`? Can you nail it to a certain packet size?

**Comment by small-5** [comment]
I used the -H parameter, I will test it later according to your suggestion and tell you the result.Thank you!

**Comment by Logan007** [comment]
Then, please also without `-H`.

**Comment by small-5** [comment]
After testing, it is found that this problem is related to the -H parameter. Enabling the -H parameter will cause packet loss and slow speed problems. This problem has nothing to do with the encryption method. And when some edge nodes enable the -H parameter, and some edge nodes do not enable the -H parameter, it will cause a segmentation fault problem on the supernode, just like my previous feedback: https://github.com/ntop/ n2n/issues/453.

**Comment by small-5** [comment]
After testing, it is found that this problem is related to the -H parameter. Enabling the -H parameter will cause packet loss and slow speed problems. This problem has nothing to do with the encryption method. And when some edge nodes enable the -H parameter, and some edge nodes do not enable the -H parameter, it will cause a segmentation fault problem on the supernode, just like my previous feedback: https://github.com/ntop/n2n/issues/453.

**Comment by small-5** [comment]
The program compiled with commit 73fa50a37dc23a91a130135583494c97a2355a95 does not have this problem with the -H parameter, so the problem should be easy to troubleshoot.

**Comment by Logan007** [comment]
Thank you for you feedback, I will have a closer look if there were changes to the `-H` part which I can't remember right now.

However, it is quite a wide range of commits. Are you able to specify the range a little further, e.g. by compiling half the way between the ones, quarter, … and so on?

Also, as I recently have seen have some issues with `-Os`, could you please compile without, e.g. using `-O3` instead?

**Comment by Logan007** [comment]
Thank you! 

I have not seen such segfault out here on x86even even though extensively using `-H`. That's what made me wonder… When exactly does the segfault happen? As soon as the second edge without `-H` connects?

I only have x86 and a Raspberry Pi to my avail and am absolutely unaware of openwrt compilation. That's why I need your support while going through the code:

`git diff 73fa50a..644ea6c` delivers huge output. For it does not remain the hard-to-find needle in the haystack, I need to ask you again to support by narrowing down towards the possibly corrupting commit by compiling and testing a few more commits in between (half way between the two mentioned commits, then quarter way, … and so on).

**Comment by small-5** [comment]
I will test it later to see if I can pinpoint which version is causing the problem.
Segmentation fault occurs when a node uses the -H parameter, and when a node is not used, it will appear if the supernode is restarted.
Another thing to say is that I changed the community.list of supernode without using the -H parameter in order to test. Only the name temporarily used for testing was allowed. The name I usually use is not in the list. The segmentation fault occurs when the supernode is restarted under such circumstances.

**Comment by small-5** [comment]
Segfault also exists in my test on ubuntu 20.04.

**Comment by small-5** [comment]
I noticed the parameter change, I now use -p instead of -l.

**Comment by Logan007** [comment]
This is so strange. I have just re-compiled even my own public supernode. I do not see no segfault… Please try to narrow the pull request that breaks it for you – my only hope!

**Comment by small-5** [comment]
After I test, solve the problem of packet loss and slow rate caused by the -H parameter. The segmentation fault may be related to this problem.

**Comment by Logan007** [comment]
> The segmentation fault occurs when the supernode is restarted under such circumstances.

So, it crashed for you if an un-encrypted packet arrives with a community which is not in the list?

**Comment by Logan007** [comment]
> So, it crashed for you if an un-encrypted packet arrives with a community which is not in the list?

Now, I understand it better and I was able to reproduce the bug. I am hunting…

**Comment by Logan007** [comment]
I found one issue with the code. The following helped me to prevent the seg fault. Could you please apply the following change to your code and check if you still have the seg fault **and** if traffic is still slow/lossy?

- `src/sn_utils.c` comment the lines 1070-1074 (the whole `if (from_supernode…` block)

**Comment by small-5** [comment]
I commented 1070-1074 lines, but it didn’t solve slow and lossy

**Comment by Logan007** [comment]
Does it fix the `-H` seg fault issues for you?

**Comment by small-5** [comment]
It seems that there is no segmentation fault for the time being,but slow and lossy still exist.

**Comment by Logan007** [comment]
Okay, that's the part where I need your help for finding the commit. I will have a close look at the current code anyway.

**Comment by small-5** [comment]
I found it, it is the problem that appeared after this version: https://github.com/ntop/n2n/commit/0776e069124728b5f62c01b6b6482660a1b38c2d. In the previous version https://github.com/ntop/n2n/commit/40f39caaac8145644a292420001a316296c9f2ed is still normal.

**Comment by small-5** [comment]
After adding replay protection,slow and lossy appeared.

**Comment by Logan007** [comment]
Thank you!!

I will have a closer look and get back to you. As I do not have your platform available, I might ask you to test one thing or another…

**Comment by Logan007** [comment]
Upfront, you could try the two following things:

- Make sure the clocks of edges and supernode all are somewhat in sync, +/- 16 seconds are allowed as a maximum value
- Your connections / platforms might require a higher time stamp jitter. Please try to change the value of`TIME_STAMP_JITTER` at line 62 of `include/n2n_define.h` from `0x0000000027100000LL` to `0x0000000271000000LL` or even higher and re-compile.

**Comment by small-5** [comment]
I now set it to 0x00000000F4240000LL (1000ms), there is no slow and lossy.

**Comment by small-5** [comment]
No need to set to 0x0000000271000000LL (2560ms), it has been solved.Maybe it can be adjusted below 1000ms, I haven't tested it yet

**Comment by small-5** [comment]
With the exception of one edge, everything else is X64. The supernode CPU model is N3710. These X64 devices have no problem under the 160ms setting. The problem is a MIPSEL device, the CPU model is MT7621. I will continue to test later to see how much delay this device can be set to.Thank you!

**Comment by small-5** [comment]
In addition, the delay between the two nodes used for testing is about 13ms.

**Comment by small-5** [comment]
The IP 10.1.1.1 device also runs on Openwrt and is also a super node. Is it still related to CPU performance?

**Comment by Logan007** [comment]
> Let me think through for 1500 ms…

If we set the limit to 1500 ms in the project and assuming that some systems lack sub-second accuracy, I would at least need 24 more bits for hashing/check-summing and preferably 12 more random bits to reasonably protect against replays on a 1 GBps connection.

While I already am having ideas on how to solve the bit-question, I fear that longer hashes will take more CPU time on edges as well as supernodes. Interesting challenge…

**Comment by small-5** [comment]
Under the setting of 1000ms, the experience of visiting the 10.1.1.2 webpage is similar to the previous version without replay attack.

**Comment by small-5** [comment]
After my test, when it is set to 1000ms, the experience of the web application is similar to the previous version without replay protection, so I think it is fine to set it to 1000ms. What I use here is ChaCha20 encryption. After my test, this time setting is not directly related to bandwidth. I use -Os to compile (I did not add additional FLAGS).

**Comment by Logan007** [comment]
> After my test, when it is set to 1000ms, the experience of the web application is similar to the previous version without replay protection, so I think it is fine to set it to 1000ms. 

Okay, I will set it to 1000 ms very soon and think about checksum bits at the same time. That's my homework I take from this Issue's discussion. After all, it remains a bit of a mystery especially as the ping of 13 ms does not indicate too many possible different routes on which packets could overtake each other. My best guess is that it is due to a lower sub-second accuracy.

> After my test, this time setting is not directly related to bandwidth. 

I see. So, your CPU probably is not saturated.

> I use -Os to compile (I did not add additional FLAGS).

Some platforms really benefit from `-march=native` if you compile locally. Even if you do not use AES as a cipher, the hashing required for header encryption significantly benefits from the AES-NI instruction set. The new `tools/n2n-benchmark` clearly shows it (now including the hash, too) – at least on my platforms, your mileage may vary.

And ChaCha20 will fly if compiled with SSE support (again, `-march=native`).

**Comment by small-5** [comment]
Thank you, I will add this parameter to compile and test according to your suggestion.

**Comment by small-5** [comment]
I also think it is caused by a lower sub-second accuracy.

**Comment by Logan007** [comment]
> Thank you, I will add this parameter to compile and test according to your suggestion.

Just one quick thought: It will not work for for cross-compiling, only if compiled on your target-platform because `-march=native` takes in the local native flags features). The Intel CPU should be able to compile locally, I guess.

Let us know the results, especially how the benchmark outputs compare (before/after).

**Comment by small-5** [comment]
I will test tomorrow to see if it can be compiled with SDK and run on the target machine.

**Comment by small-5** [comment]
After adding -march=native, it cannot run on the target machine, so there is no way to compare. The following figure is a screenshot of the test on the N3710 after compiling without the -march=native parameter.
![1](https://user-images.githubusercontent.com/61473216/96032596-791e2e80-0e91-11eb-959b-c44eee7a6070.jpg)

**Comment by small-5** [comment]
I try to compile directly on the openwrt platform to see what happens.

**Comment by small-5** [comment]
Unfortunately, the firmware partition size I set is not enough to install gcc make autoconf automake, so I can only test the partition when I compile the firmware next time.

**Comment by small-5** [comment]
Unfortunately, the firmware partition size I set is not enough to install gcc make autoconf automake, so I have to adjust the partition size the next time I compile the firmware before testing.

**Comment by Logan007** [comment]
> Unfortunately, the firmware partition size I set is not enough to install gcc make autoconf automake, so I have to adjust the partition size the next time I compile the firmware before testing.

There is another way. If cross-compiling, you could specify the machine-specific flags directly. For your Intel CPU, the major ones should be `-maes -mssse3 -mrdrnd -O3` There might be some more of them for special CPU features.

**Comment by small-5** [comment]
With -march=silvermont -O3:
![-march=silvermont -O3](https://user-images.githubusercontent.com/61473216/96117040-06a66080-0f1c-11eb-8598-cdfc2cdf97af.jpg)
With -maes -mssse3 -mrdrnd -O3:
![-maes -mssse3 -mrdrnd -O3](https://user-images.githubusercontent.com/61473216/96117062-102fc880-0f1c-11eb-9e30-60cd9d564022.jpg)
The performance of cc20 and speck has dropped, which is strange.

**Comment by Logan007** [comment]
Indeed, that is strange! SSE is supposed to fasten things (and it does, at least on my platform)… But AES looks great, also the hashing gained.

**Comment by small-5** [comment]
Performance degradation may be related to Openwrt.

**Comment by Logan007** [comment]
> Performance degradation may be related to Openwrt.

I do not think so. Another OS will not be able to do better on the same platform.

It might be about platform-dependent factors like RAM speed, cache size, bus width, … as well as the code which harmonizes more with one platform than with another.

Wow, what a long discussion – a great one. I have already opened a new issue to work on header encryption. Will take a while. I might ask for testing once it is done because your setup seems to be relevant in that respect. In the meanwhile, you could go with the other time jitter value, I guess.

Again, thank you for actively supporting the development of this project!

**Comment by small-5** [comment]
As far as my current equipment is concerned, using 1000ms is the smallest setting value that does not affect the experience.I believe others will encounter similar situations.

**Comment by Logan007** [comment]
I understood that it especially affects the MIPSEL, right?

I might think of something that could also deal with less sub-second accuracy, maybe counter style for sub-second range. Dooes using `-vvvvv` on edge and supernode show "due to time stamp error" messages?

**Comment by small-5** [comment]
I haven't used debug to check the error message. I will test it when I have time tomorrow. But now using 1000ms, this error message may not appear.

**Comment by small-5** [comment]
I don't know whether this problem is caused by Openwrt or MIPSEL, but I feel that it is caused by the sub-second accuracy of Openwrt.

**Comment by small-5** [comment]
> I might think of something that could also deal with less sub-second accuracy, maybe counter style for sub-second range. Does using `-vvvvv` on edge and supernode show "due to time stamp error" messages?

The supernode I compiled with the 160ms setting runs on openwrt, and the following 2 errors will be reported:
20/Oct/2020 00:56:10 [n2n.c:495] time_stamp_verify_and_update found a timestamp too old compared to previous.
20/Oct/2020 00:56:10 [sn_utils.c:912] process_udp dropped PACKET due to time stamp error.
I run edge on the supernode device, and the address used is 127.0.0.1 instead of the router's external network address. In this case, the connection established with other edges is not P2P, so this situation occurs.

**Comment by small-5** [comment]
> I might think of something that could also deal with less sub-second accuracy, maybe counter style for sub-second range. Does using `-vvvvv` on edge and supernode show "due to time stamp error" messages?

The supernode I compiled with the 160ms setting runs on openwrt, and the following 2 errors will be reported:
20/Oct/2020 00:56:10 [n2n.c:495] time_stamp_verify_and_update found a timestamp too old compared to previous.
20/Oct/2020 00:56:10 [sn_utils.c:912] process_udp dropped PACKET due to time stamp error.
I run edge on the supernode device, and the address used is 127.0.0.1 instead of the router's external network address. In this case, the connection established with other edges is not P2P, so this situation occurs.
If I use the router’s external network address, I can establish a P2P connection with other edge, and then there will be no time stamp error.
So please take a look at how to deal with this problem.Thank you!

**Comment by small-5** [comment]
In addition, if the supernode is compiled with 1000ms settings, these two errors will not appear in the case of non-P2P.

**Comment by Logan007** [comment]
Thank you for reporting back! 

That is indeed interesting… especially as the error occurs at the supernode and with local address (which should show less delay). I will try to further drill down to it.

**Comment by small-5** [comment]
In the case of using 127.0.0.1 to connect to the supernode, it cannot be connected via P2P. I wonder if this can be solved.

**Comment by Logan007** [comment]
> In the case of using 127.0.0.1 to connect to the supernode, it cannot be connected via P2P. I wonder if this can be solved.

This is the point at which my network skills are about to exhaust but my best guess is that this is **not** possible because the supernode only sees the local address as "public face" of the edge in this case. If it distributes it among the other edges, they will not be able to connect.

I am still not sure how and if this relates to the time stamp error.

**Comment by small-5** [comment]
The timestamp error only occurs in the case of non-p2p connection. In fact, the main problem is to solve this problem. It is not very important whether the P2P connection can be established by using 127.0.0.1 to connect supernode

**Comment by small-5** [comment]
There are some errors in the previous feedback. I just tested and found that if the edge is compiled with a 160ms setting, this error will occur even if it is a P2P connection.

**Comment by Logan007** [comment]
This observation makes the things seem to be more in line… As we ruled out CPU saturation and other things, I stick with openWrt's lower sub-second accuracy.

**Comment by small-5** [comment]
Yes,this problem is definitely caused by openwrt.I compiled a edge on the windows platform. Under the condition of 160ms, even if the P2P connection is made,this error will still be reported.

**Comment by Logan007** [comment]
@small-5  May I ask you to try to compile this program and run it a few times on you wrt box via ssh or so and share the output? I suspect that only `date`  is affected by the accuracy issue, so maybe I can fix it by changing the way the time stamp is generated (so far we have been using `gettimeofday()`).

```C

#include <stdio.h>
#include <time.h>

int main(int argc, char **argv, char **arge) {

  struct timespec t;

  clock_gettime(CLOCK_REALTIME, &t);
  printf("     start: %lu s %lu ns\n", t.tv_sec,  t.tv_nsec);

  clock_getres(CLOCK_REALTIME, &t);
  printf("resolution: %lu s %lu ns\n", t.tv_sec,  t.tv_nsec);

  clock_gettime(CLOCK_REALTIME, &t);
  printf("      stop: %lu s %lu ns\n", t.tv_sec,  t.tv_nsec);
}
```

Thank you!

**Comment by small-5** [comment]
@Logan007 I'll test it tomorrow and take a screenshot. Thank you!

**Comment by small-5** [comment]
![image](https://user-images.githubusercontent.com/61473216/100070224-a25cb200-2e74-11eb-8032-b62ca7ea7f50.png)
In the previous problem on the device to compile this code run results, please see if there is a solution.Thank you!

**Comment by Logan007** [comment]
This indeed is very helpful, thank you so much!!!

It shows that we can use `clock_gettime()` for time stamp generation (instead of `gettimeofday()`) as it allows sufficient accuracy required for the "regular" timestamp jitter (no need to use 1 second time frames then :relieved: ).

Starting from here, with this information, I already am reworking the time stamp format a bit and will apply some other strengthening changes to header encryption. I estimate that it will take up to two weeks to have it all finished.

**Comment by small-5** [comment]
Thank you!

**Comment by Logan007** [comment]
So, I changed the timer source with PR #522. I works fine on my x86 Linux boxes and my best guess is that it also should solve this issue without having to change the jitter time frame to one second anymore. You could pre-test – if you dare to do so, feedback would be appreciated.

However, someone would have to test it on Windows (Mac should be fine from what I have read) before it could safely be merged into dev.

Once #522 proves to work well, I will bring the other changes (#466) to header encryption.

**Comment by small-5** [comment]
@Logan007 Thank you. I'll let you know the result after the test

**Comment by small-5** [comment]
@Logan007 I use the source code of your warehouse to compile and test, it is still the same situation.

**Comment by Logan007** [comment]
First of all: Thank you for testing!

So, same situation means that it works with a jitter of 1000 ms?

Did you replace the code on all edges and also the supernode?

**Comment by small-5** [comment]
I have recompiled all edge with the source code of your repositorie, and supernode is also recompiled. In addition, I found that the bin compiled on openwrt is the same size as the program generated by e66f913.So I feel that the changed part has not taken effect.Please handle it, thank you!

**Comment by Logan007** [comment]
> I have recompiled all edge with the source code of your repositorie, and supernode is also recompiled. In addition, I found that the bin compiled on openwrt is the same size as the program generated by e66f913.

Again, thank you for your feedback.

> So I feel that the changed part has not taken effect.

I am getting more and more clueless about that. As I have not heard any similar report so far, maybe we need to shift our view and look at another part maybe triggering this behavior. So, for now, your best option seemingly is to use that 1 second limit.

> Please handle it, thank you!

I am reworking header encryption anyway, maybe it will help to solve this issue.

**Comment by small-5** [comment]
Thank you!

**Comment by Logan007** [comment]
@small-5 do you want to try #595? You would need to put this version on all your edges and supernodes.

If this one does not help your scenario, I do not know what else could...

**Comment by Logan007** [comment]
Turns out there was a bug in #595 which is fixed by now. The code already made it to current _dev_.

**Comment by Logan007** [comment]
I assume #595 along with #612 have solved it. If not, please re-open.
